File: Assets/Scripts/Heights Scene/FallingBehavior.lua
local CS = CS
local UnityEngine = CS.UnityEngine
local Vector3 = UnityEngine.Vector3
local Time = UnityEngine.Time
local coroutine = coroutine

local isFalling = false
local fallDistance = 20
local fallDuration = 2


function awake()
    if xrOrigin == nil then
        print("xrOrigin not assigned in Lua!")
    end

    if gameObject == nil then
        print("gameObject is nil!")
    else
        triggerObject = gameObject:GetComponent(typeof(UnityEngine.Collider))
        print("Trigger object:", triggerObject)
    end
end

function on_trigger_enter(other)
    if not isFalling and other.gameObject.tag == "MainCamera" then
        isFalling = true
        elapsed = 0
        startPos = xrOrigin.transform.position
        endPos = startPos + Vector3.down * fallDistance
    end
end

function update()
    if isFalling then
        if elapsed < fallDuration then
            local t = elapsed / fallDuration
            xrOrigin.transform.position = Vector3.Lerp(startPos, endPos, t)
            elapsed = elapsed + Time.deltaTime
        else
            xrOrigin.transform.position = Vector3(0, 1, 0)
            isFalling = false
        end
    end
end

File: Assets/Scripts/Heights Scene/FallingBehavior.cs
using UnityEngine;
using System.Collections;

public class FallingBehavior : MonoBehaviour
{
    [SerializeField] private Transform xrOrigin;
    [SerializeField] private float fallDistance = 20f;
    [SerializeField] private float fallDuration = 2f;
    // [SerializeField] private AudioClip fallSound;
    // [SerializeField] private AudioSource audioSource;

    private bool isFalling = false;

    void  OnTriggerEnter(Collider other)
    {
        if (other.gameObject.CompareTag("MainCamera") && !isFalling)
        {
            StartCoroutine(FallRoutine());
            Debug.Log("Falling triggered!");
        }

    }

    IEnumerator FallRoutine()
    {
        isFalling = true;

        // if (fallSound && audioSource)
        //     audioSource.PlayOneShot(fallSound);

        Vector3 start = xrOrigin.position;
        Vector3 end = start + Vector3.down * fallDistance;

        float elapsed = 0f;
        while (elapsed < fallDuration)
        {
            float t = elapsed / fallDuration;
            xrOrigin.position = Vector3.Lerp(start, end, t);
            elapsed += Time.deltaTime;
            yield return null;
        }

        xrOrigin.position = new Vector3(0f, 1f, 0f); 
        isFalling = false;
    }
}